src/base/SolutionArray.cpp

```diff
@@ -36,30 +36,229 @@ const std::map<std::string, std::string> aliasMap = {
 namespace Cantera
 {
 
-SolutionArray::SolutionArray(
-    const shared_ptr<Solution>& sol,
-    size_t size,
-    const AnyMap& meta)
+SolutionArray::SolutionArray(const shared_ptr<Solution>& sol,
+                             size_t size, const AnyMap& meta)
     : m_sol(sol)
     , m_size(size)
+    , m_dataSize(size)
     , m_meta(meta)
 {
     if (!m_sol) {
         throw CanteraError("SolutionArray::SolutionArray",
             "Unable to create SolutionArray from invalid Solution object.");
     }
     m_stride = m_sol->thermo()->stateSize();
-    m_data.resize(m_size * m_stride, 0.);
+    m_data.reset(new vector<double>(m_dataSize * m_stride, 0.));
+    m_extra.reset(new map<string, AnyValue>());
+    m_order.reset(new map<int, string>());
+    for (size_t i = 0; i < m_dataSize; ++i) {
+        m_active.push_back(i);
+    }
+    reset();
+    m_apiShape.resize(1);
+    m_apiShape[0] = m_dataSize;
+}
+
+SolutionArray::SolutionArray(const shared_ptr<SolutionArray>& other,
+                             const vector<int>& selected)
+    : m_sol(other->solution())
+    , m_size(selected.size())
+    , m_dataSize(other->m_data->size())
+    , m_stride(other->m_stride)
+    , m_data(other->m_data)
+    , m_extra(other->m_extra)
+    , m_order(other->m_order)
+    , m_shared(true)
+    , m_active(selected)
+{
+    for (auto loc : m_active) {
+        if (loc < 0 || loc >= m_dataSize) {
+            IndexError("SolutionArray::SolutionArray", "indices", loc, m_dataSize);
+        }
+    }
+    set<int> unique(selected.begin(), selected.end());
+    if (unique.size() < selected.size()) {
+        throw CanteraError("SolutionArray::SolutionArray", "Indices must be unique.");
+    }
+}
+
+template<class T>
+void __resetSingle(AnyValue& extra, const vector<int>& slice);
+
+template<class T>
+void __resetMulti(AnyValue& extra, const vector<int>& slice);
+
+void SolutionArray::reset()
+{
+    size_t nState = m_sol->thermo()->stateSize();
+    vector<double> state(nState);
+    m_sol->thermo()->saveState(state); // thermo contains current state
+    for (int k = 0; k < m_size; ++k) {
+        std::copy(state.begin(), state.end(), &(*m_data)[m_active[k] * m_stride]);
```

Comment:
An alternative that might be slightly less clunky (here and in several other locations):
```suggestion
        std::copy(state.begin(), state.end(), m_data->data() + m_active[k] * m_stride);
```